Nullable.md

```diff
@@ -0,0 +1,191 @@
+`Union{Some{Void}, Void}` -- as might happen in a hash table implementation
+that returns `Some` when a key is present and associated with `nothing` as a
+value, but returns `Void` when a key is not present.
+
+### `Union{T, Null{T}}` Pros & Cons
+
+This approach allows us to distinguish `Null{Bool}` from `Null{Int}` in
+cases in which that distinction might be useful. R makes such a distinction,
+although it seems to be done only because it is falls out naturally from an
+implementation that uses sentinel values. Otherwise, it is similar to the
+`Union{T, Void}` case described above, except for one crucial problem.
+
+That problem is essentially the problem we already have with `Nullable{T}`:
+writing "clean" functions requires information about type-inference because of
+a coupling of result types to knowledge about counterfactual outcomes in
+branches.
+
+To see the issue, consider the following possible implementation of
+`map(f, x::Nullable{T})` in our current implementation of `Nullable{T}`:
+
+```
+function map(f, x::Nullable{T})
+    if hasvalue(x)
+        return Nullable(f(value(x)))
+    else
+        return Nullable{return_types(f, T)}()
+    end
+end
+```
+
+The problem is that, when given a null input, the return type can only match
+the counterfactual return type that would have occurred if called on a non-null
+value if the function knows the return type of `f`.
+
+In static languages with arrow types this is not a problem as the return type
+of a function is guaranteed to be well-defined by the semantics of the
+language. But Julia does not provide such guarantees, so type-inference must
+be invoked to determine information about the return type of the counterfactual
+branch.
+
+Note that this problem does not occur for the non-null case because we are
+expected to evaluate `f`. It is only a problem when we need to know the return
+type of `f`, but must not determine this type by evaluation of `f`.
+
+## Step 2:: Define Core Operations on Nullable Types by Lifting
+
+The core semantic issue for nullable types is maximizing code reuse: we want
+to evaluate `f(x::?T)`, but not require that every function be defined for
+both `f(x::T)` and `f(x::?T)` separately as the overwhelming majority of
+functions will return `null` as their output whenever any of their arguments
+are `null`. Implementing this lifting of functions can be solved as follows:
+
+First, we define `broadcast(f, x::Nullable)` as follows:
+
+```
+immutable Lifted{T}
+    func::T
+end
+
+lift(f) = Lifted(f)
+
+broadcast(f, x::Nullable) = broadcast(lift(f), x)
+
+broadcast(f′::Lifted, x::Nullable) = hasvalue(x) ? f′.func(value(x)) : null
+```
+
+This allows us to give default semantics to all functions that implement the
+natural approach to lifting, while still making it possible to handle special
+case functions that require distinct semantics.
+
+Some such functions are:
+
+* `hasvalue`
+* `isnull`
+* `get`
+* `get_or_default`
```

Comment:
I did this for C# comparisons. This is our two argument `get`. Will indicate that.